packages/core/src/typeschema/validation.ts

```diff
@@ -100,54 +102,57 @@ class ResourceValidator {
     }
   }
 
-  private validateObject(value: TypedValue, schema: InternalTypeSchema, path: string): void {
+  private validateObject(obj: TypedValue, schema: InternalTypeSchema, path: string): void {
     for (const [key, _propSchema] of Object.entries(schema.fields)) {
-      this.checkProperty(value, key, schema, path + '.' + key);
+      this.checkProperty(obj, key, schema, `${path}.${key}`);
     }
 
     //@TODO(mattwiller 2023-06-05): Detect extraneous properties in a single pass by keeping track of all keys that
     // were correctly matched to resource properties as elements are validated above
-    this.checkAdditionalProperties(value, schema.fields, path);
+    this.checkAdditionalProperties(obj, schema.fields, path);
   }
 
-  private checkProperty(value: TypedValue, key: string, schema: InternalTypeSchema, path: string): void {
-    const propertyValues = getNestedProperty(value, key);
+  private checkProperty(parent: TypedValue, key: string, schema: InternalTypeSchema, path: string): void {
+    const propertyValues = getNestedProperty(parent, key);
     const element = schema.fields[key];
     if (!element) {
       throw new Error(`Missing element validation schema for ${key}`);
     }
-    for (const property of propertyValues) {
-      if (isEmpty(property)) {
+    for (const value of propertyValues) {
+      if (value === undefined) {
         if (element.min > 0) {
           this.issues.push(createStructureIssue(path, 'Missing required property'));
         }
         return;
+      } else if (isEmpty(value)) {
+        this.issues.push(createStructureIssue(path, 'Invalid empty value'));
+        return;
       }
 
       let values: TypedValue[];
       if (element.isArray) {
-        if (!Array.isArray(property)) {
-          this.issues.push(createStructureIssue(path, 'Expected array of values'));
+        if (!Array.isArray(value)) {
+          this.issues.push(createStructureIssue(path, 'Expected array of values for property'));
           return;
         }
-        values = property;
+        values = value;
       } else {
-        if (Array.isArray(property)) {
-          this.issues.push(createStructureIssue(path, 'Expected single value'));
+        if (Array.isArray(value)) {
+          this.issues.push(createStructureIssue(path, 'Expected single value for property'));
           return;
         }
-        values = [property as TypedValue];
+        values = [value as TypedValue];
       }
 
       if (element.pattern) {
-        if (!deepIncludes(element.pattern, property)) {
+        if (!deepIncludes(value, element.pattern)) {
```

Comment:
FYI @jamestouri I swapped the order of the params here while resolving a merge conflict